### PR TITLE
Number of users online

### PR DESCRIPTION
**Pull request by parenthetical**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -53,6 +54,7 @@ export default function ChatApp() {
         width: "100%",
       }}
     >
+      <Presence />
```
Comment (stevekrouse) [inline src/App.tsx]: Probably put this in the Header component itself, and on the right corner of the screen

File: src/components/Presence.tsx [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { usePresence } from "../state/presence";
+
+export default function Presence() {
+  const numberOnline = usePresence();
+  return <div>{numberOnline} users active.</div>;
```
Comment (stevekrouse) [inline src/components/Presence.tsx]:
```suggestion
  return <div>{numberOnline} online now</div>;
```

File: src/state/presence.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      dispatchPresence(nw);
+      console.log(presenceDb);
+      setNumberOnline(
+        Object.values(presenceDb || {}).filter(
+          (lastSeen: number) => lastSeen > nw - 30 * 1000
+        ).length
+      );
+    }, 1000 * 20);
+  });
+  return numberOnline;
```
Comment (stevekrouse) [inline src/state/presence.ts]:
```suggestion
  useEffect(() => {
    // dispatch presence on first, and every 20 seconds thereafter
    dispatchPresence(Date.now());
    const interval = setInterval(() =>
      dispatchPresence(Date.now());        
    , 1000 * 20);
    return () => clearInterval(interval)
  }, [userId]); // do this effect when the userId changes
  return Object.values(presenceDb).filter(
      (lastSeen: number) => lastSeen > nw - 30 * 1000
    ).length
  );
```

Comment (stevekrouse) [reply src/state/presence.ts]: Do you see why we calculate the `numberOnline` value outside of the interval? (It's a reactive value. Hooks are weird!)

The code I wrote above requires that you pull in the `useUser` hook in this function

File: src/state/presence.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+export const usePresence = () => {
+  const [presenceDb, dispatchPresence] = useCloudReducer<
+    PresenceDB,
+    PresenceAction,
+    PresenceActionError
+  >({
+    name: `${appName}/user-presence`,
+    initialState: {},
+    reducer: (users, lastSeen, { userId }) => {
+      users[userId] = lastSeen;
```
Comment (stevekrouse) [inline src/state/presence.ts]:
```suggestion
      if (usersId) {
        users[userId] = lastSeen;
      }
```

File: src/state/presence.ts (hunk 3 above)
Comment (parenthetical) [reply src/state/presence.ts]: I think I get it!

**Pull request merged**